modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManagerWithTieredStorage.java

```diff
@@ -0,0 +1,412 @@
+  private boolean isRemoteAssetStoreId(String storeId) {
+    return !isLocalAssetStoreId(storeId);
+  }
+
+  @Override
+  public Opt<Asset> getAsset(Version version, final String mpId, final String mpeId) {
+    // try to fetch the asset
+    for (final AssetDtos.Medium asset : getDb().getAsset(RuntimeTypes.convert(version), mpId, mpeId)) {
+      for (final String storageId : getSnapshotStorageLocation(version, mpId)) {
+        for (final AssetStore store : getAssetStore(storageId)) {
+          for (final InputStream assetStream : store.get(StoragePath.mk(asset.getOrganizationId(), mpId, version, mpeId))) {
+            final Asset a = new AssetImpl(
+                    AssetId.mk(version, mpId, mpeId),
+                    assetStream,
+                    asset.getAssetDto().getMimeType(),
+                    asset.getAssetDto().getSize(),
+                    asset.getStorageId(),
+                    asset.getAvailability());
+            return Opt.some(a);
+          }
+        }
+      }
+    }
+    return Opt.none();
+  }
+
+  /** Move the assets for a snapshot to the target store */
+  private void copyAssetsToStore(Snapshot snap, AssetStore store) throws Exception {
+    final String mpId = snap.getMediaPackage().getIdentifier().toString();
+    final String orgId = snap.getOrganizationId();
+    final Version version = snap.getVersion();
+    final String prettyMpId = mpId + "@v" + version;
+    logger.debug(format("Moving assets for snapshot %s to store %s", prettyMpId, store.getStoreType()));
+    for (final MediaPackageElement e : snap.getMediaPackage().getElements()) {
+      if (!MOVABLE_TYPES.contains(e.getElementType())) {
+        logger.debug(format("Skipping %s because type is %s", e.getIdentifier(), e.getElementType()));
+        continue;
+      }
+      logger.debug(format("Moving %s to store %s", e.getIdentifier(), store.getStoreType()));
+      final StoragePath storagePath = StoragePath.mk(orgId, mpId, version, e.getIdentifier());
+      final Opt<StoragePath> existingAssetOpt = findAssetInVersionsAndStores(e.getChecksum().toString(), store.getStoreType());
+      if (existingAssetOpt.isSome()) {
+        final StoragePath existingAsset = existingAssetOpt.get();
+        logger.debug("Content of asset {} with checksum {} already exists in {}",
+                existingAsset.getMediaPackageElementId(), e.getChecksum(), store.getStoreType());
+        if (!store.copy(existingAsset, storagePath)) {
+          throw new AssetManagerException(
+                  format("An asset with checksum %s has already been archived but trying to copy or link asset %s to it failed",
+                          e.getChecksum(), existingAsset));
+        }
+      } else {
+        final Opt<Long> size = e.getSize() > 0 ? Opt.some(e.getSize()) : Opt.<Long>none();
+        store.put(storagePath, Source.mk(e.getURI(), size, Opt.nul(e.getMimeType())));
+      }
+    }
```

Comment:
Why are you not moving the manifest.xml too? Is the manifest.xml an exact copy of what is in the database?

Reply:
The manifest isn't included in the MP afaik.  It also does not get updated beyond the initial ingest, so it would be out of date regardless.

Reply:
The manifest.xml is a copy of the media package for that version. When a snapshot is taken, it's created in the /archive file system. It seems to be an exact copy of the mediapackage_xml column in the oc_assets_snapshot table. So it doesn't need to be copied? Can you double-check if this is true? I am just afraid of deleting something that may be needed later...